patroni/utils.py

```diff
@@ -1101,24 +1157,43 @@ def keepalive_socket_options(timeout: int, idle: int, cnt: int = 3) -> Iterator[
                 * ``TCP_KEEPCNT``.
     """
     yield (socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
-
-    if not (sys.platform.startswith('linux') or sys.platform.startswith('darwin')):
-        return
-
-    if sys.platform.startswith('linux'):
-        yield (socket.SOL_TCP, 18, int(timeout * 1000))  # TCP_USER_TIMEOUT
-
+    TCP_USER_TIMEOUT = getattr(socket, 'TCP_USER_TIMEOUT', None)
+    if TCP_USER_TIMEOUT is not None:
+        timeout_in_milliseconds = int(timeout * 1000)
+        max_possible_value = max_keepalive_option_value("TCP_USER_TIMEOUT")
+        if max_possible_value is not None and timeout_in_milliseconds > max_possible_value:
+            max_possible_value_in_seconds = int(max_possible_value / 1000)
+            logger.warning(
+                "TCP_USER_TIMEOUT must not exceed {0} milliseconds, so it is changed from {1} seconds to {2} seconds."
+                .format(max_possible_value, timeout, max_possible_value_in_seconds)
+            )
+            timeout_in_milliseconds = max_possible_value
+        yield (socket.SOL_TCP, TCP_USER_TIMEOUT, timeout_in_milliseconds)
     # The socket constants from MacOS netinet/tcp.h are not exported by python's
     # socket module, therefore we are using 0x10, 0x101, 0x102 constants.
     TCP_KEEPIDLE = getattr(socket, 'TCP_KEEPIDLE', 0x10 if sys.platform.startswith('darwin') else None)
     if TCP_KEEPIDLE is not None:
+        max_possible_value = max_keepalive_option_value("TCP_KEEPIDLE")
+        if max_possible_value is not None and idle > max_possible_value:
+            logger.warning(
+                "TCP_KEEPIDLE must not exceed {0}, so it is changed from {1} to {0}."
+                .format(max_possible_value, idle)
+            )
+            idle = max_possible_value
         yield (socket.IPPROTO_TCP, TCP_KEEPIDLE, idle)
     TCP_KEEPINTVL = getattr(socket, 'TCP_KEEPINTVL', 0x101 if sys.platform.startswith('darwin') else None)
     if TCP_KEEPINTVL is not None:
         intvl = keepalive_intvl(timeout, idle, cnt)
         yield (socket.IPPROTO_TCP, TCP_KEEPINTVL, intvl)
     TCP_KEEPCNT = getattr(socket, 'TCP_KEEPCNT', 0x102 if sys.platform.startswith('darwin') else None)
     if TCP_KEEPCNT is not None:
+        max_possible_value = max_keepalive_option_value("TCP_KEEPCNT")
+        if max_possible_value is not None and cnt > max_possible_value:
+            logger.warning(
+                "TCP_KEEPCNT must not exceed {0}, so it is changed from {1} to {0}."
+                .format(max_possible_value, cnt)
+            )
+            cnt = max_possible_value
```

Comment:
It all looks like overenginieering to me.
Why not just have a function `apply_keepalive_limit(option: str, value: int) -> int` which will make sure that the limit is not exceeded, or show the warning and adjust it.

Another question is whether we need to have a warning. IMO it could be a debug.

And the third problem - we prefer to use "%s" placeholders when working with logger instead of str.format or f-string.

Reply:
Thank you for your opinions; they are great. I pushed another commit and solved the problems.